### DISCUSSION TITLE
Reusing a same set of dsl-rulesets

### DISCUSSION THREAD

**Discussion by franckvallee** [evrete/evrete#16] [General]
Hello, 

I tried to declare a java @ruleset file which contains a class that inherits from another ruleset, but I got an error.
Do you think it is possible?

java CommonRS.java
>@Ruleset("common-rs")
class CommonRS {
    @Rule(A)
    public void reusableRule() { ... }
}

java BusinessRS.java
>@Rueset("Business")
class BusinessRS extends CommonRS {
 @Rule(B) 
 public void aBusinessRule() { ... }
}

**Comment by andbi** [comment]
For the `JAVA-SOURCE` Dsl it is possible as long as the parent class is on the classpath or explicitly imported via the `addImport` call.

**Comment by franckvallee** [comment]
I did an explicit addImport to the knowledge object and then when I try using inheritance, I've an error:
It works fine with  @FieldDeclaration only.
But when I tried to add a @Rule in the ancestor file, I have this error:
java.lang.IllegalArgumentException: There's no declared reference '$x' in provided context.

**Reply by andbi** [reply]
The exception indicates that a `@Fact` annotation is missing on some of the method's arguments.

**Reply by franckvallee** [reply]
Yes, this resolved the point.
Now using @Fact in the ancestor makes reusable part of DSL by inheritance working.

**Reply by andbi** [reply]
The reason why you don'd need the `@Fact` annotation for `JAVA-SOURCE` rulesets, is because the engine compiles Java sources with the `-parameters` flag which preserves parameter names in the compiled byte-code. If your codebase is also compiled with this flag, then you can drop the `@Fact` annotations altogether.
